File: leetcode_python/swap_nodes_in_pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        if head is None or head.next is None:
            return head
        else:
            one, two, three = head, head.next, head.next.next
            head = two
            two.next = one # relink
        while True:
            if three is None or three.next is None:
                one.next = three
                return head
            else:
                four, five = three.next, three.next.next
                one.next = four # relink
                four.next = three # relink
                one, three = three, five


            

File: leetcode_cpp/swap_nodes_in_pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        if (!head || !head -> next) {
            return head;
        }

        ListNode *one = head, *two = head -> next, *three = head -> next -> next;
        head = two;
        two -> next = one;

        while (true) {
            if (!three || !three -> next) {
                one -> next = three;
                return head;
            }
            else {
                ListNode *four = three -> next, *five = three -> next -> next;
                one -> next = four;
                four -> next = three;
                one = three;
                three = five;
            }
        }
    }
};
